### PR TITLE
Lots of std::shared_ptr

### LINKED ISSUE #105: Use `std::shared_ptr` in TensorOperators

### PR DESCRIPTION
**Pull request by stigrj**
Extracted some of the noise from #212 which is unrelated to the input and driver.

Changes made
------------------
- Move orbitals and Fock matrix into the `Molecule` class
- Remove all `init`s of molecular properties, now they are allocated on the fly
- Introduce `std::shared_ptr` into QM operators, closes #105 
- Remove a lot of unnecessary state in the SCF solvers
- Add Fock matrix to argument list of `localize` and `orthonormalize` to make them consistent with `diagonalize`
- Update integration test references for calculations involving the `EnergyOptimizer` and `LinearResponseSolver`, since minor algorithmic changes made the results slightly different
- Adapt `SCFDriver` to all changes

Please ignore all changes in `SCFDriver`, this will be fully revised in #212.

### CODE REVIEW

File: src/SCFDriver.h [hunk 1]
```diff
@@ -205,53 +206,39 @@ class SCFDriver final {
...
 
     // XCFun
-    mrdft::XCFunctional *xcfun;
+    std::shared_ptr<mrdft::XCFunctional> xcfun{nullptr};
 
     // Perturbation operators
-    H_E_dip *h_E;  // dH/dE
-    H_B_dip *h_B;  // dH/dB
-    H_M_pso **h_M; // dH/dM[K]
+    H_E_dip *h_E{nullptr};  // dH/dE
```
Comment (robertodr) [inline src/SCFDriver.h]: Must these be pointers?

File: src/qmoperators/one_electron/DistanceOperator.h [hunk 2]
```diff
@@ -21,13 +21,13 @@ class DistancePotential final : public QMPotential {
 class DistanceOperator final : public RankZeroTensorOperator {
 public:
     DistanceOperator(double pow, const mrcpp::Coord<3> &R, double S = 1.0e-7)
-            : r_pow(pow, R, S) {
+            : r_pow(new DistancePotential(pow, R, S)) {
```
Comment (robertodr) [inline src/qmoperators/one_electron/DistanceOperator.h]: There might be copies incurred in doing the initialization like this. Probably can be moved to the body of the CTOR to use `make_shared` instead.

File: src/SCFDriver.h (hunk 1 above)
Comment (stigrj) [reply src/SCFDriver.h]: No, but it will be fixed later

File: src/qmoperators/one_electron/DistanceOperator.h (hunk 2 above)
Comment (stigrj) [reply src/qmoperators/one_electron/DistanceOperator.h]: Hmm, really? I will change to `make_shared` then

**Comment by stigrj** [comment]
I will run some tests on Fram, but at this abstraction level I doubt we'll see any change in performance

**Comment by stigrj** [comment]
No discernible change in performance

**Pull request merged**